File: java/FlattenBinaryTreetoLinkedList.java
import java.util.Stack;

/**
 * Created by Carl Shen on 1/26/15.
 * Flatten Binary Tree to Linked List
 * https://oj.leetcode.com/problems/flatten-binary-tree-to-linked-list/
 */
public class FlattenBinaryTreetoLinkedList {
    public void flatten(TreeNode root) {
        if (root == null) {
            return;
        }

        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            if (node.right != null) {
                stack.push(node.right);
            }
            if (node.left != null) {
                stack.push(node.left);
            }
            node.left = null;
            if (!stack.isEmpty()) {
                node.right = stack.peek();
            }
        }
    }

    public void flattenInplace(TreeNode root) {
        if (root == null) {
            return;
        }

        TreeNode curr = root;
        while (curr != null) {
            if (curr.left != null) {
                if (curr.right != null) {
                    TreeNode next = curr.left;
                    while (next.right != null) {
                        next = next.right;
                    }
                    next.right = curr.right;
                }
                curr.right = curr.left;
                curr.left = null;
            }
            curr = curr.right;
        }
    }

    public void flattenRec(TreeNode root) {
        if (root == null) return;
        TreeNode left = root.left, right = root.right;
        root.left = null;
        flattenRec(left);
        flattenRec(right);
        root.right = left;

        TreeNode curr = root;
        while (curr.right != null) curr = curr.right;
        curr.right = right;
    }
}

File: python/flatten_binary_tree_to_linked_list.py
#!/usr/bin/env python

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):
        if not root:
            return
        nodes = self.preorder(root)
        curr, nodes = nodes[0], nodes[1:]
        # deal with the root node
        curr.left = None
        for node in nodes:
            node.left = node.right = None
            curr.right = node
            curr = curr.right

    def preorder(self, root):
        if not root:
            return []
        result = []
        result.append(root)
        result += self.preorder(root.left)
        result += self.preorder(root.right)
        return result

    def flatten(self, root):
        if not root: return # base case
        
        self.flatten(root.left)
        self.flatten(root.right)
        
        if not root.left: return
        
        lft = root.left
        while lft.right:
            lft = root.right
        lft.right = root.left
        root.left = None
        
    def flatten(self, root):
        if not root:
            return
        
        stack = [root]
        while stack:
            node = stack.pop()
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
            
            node.left = None
            if stack: node.right = stack[-1]